feat: add compliance to API report files and its endpoint

AdriiiPRodri opened the pull request:
### Description

In this PR we have added the compliance output to the ZIP of the API outputs.

We have also added an endpoint to be able to retrieve a specific compliance within a scan, whether the results are uploaded to S3 or locally.
![image](https://github.com/user-attachments/assets/1f684706-89e5-41ab-b8c5-2ee4009cd73e)

### Checklist

- Are there new checks included in this PR? Yes / No
    - If so, do we need to update permissions for the provider? Please review this carefully.
- [x] Review if the code is being covered by tests.
- [x] Review if code is being documented following this specification https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
- [x] Review if backport is needed.
- [x] Review if is needed to change the [Readme.md](https://github.com/prowler-cloud/prowler/blob/master/README.md)
- [x] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/prowler/CHANGELOG.md), if applicable.

#### API
- [x] Verify if API specs need to be regenerated.
- [x] Check if version updates are required (e.g., specs, Poetry, etc.).
- [x] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/api/CHANGELOG.md), if applicable.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

AdriiiPRodri committed:
feat: add compliance to API report files

AdriiiPRodri committed:
feat: add download compliance endpoint

AdriiiPRodri committed:
ref: refactor generate_outputs task

AdriiiPRodri committed:
test: add unittests

AdriiiPRodri committed:
chore: restore comment

AdriiiPRodri committed:
Merge remote-tracking branch 'origin/master' into PRWLR-6373-Implement-Compliance-Outputs

AdriiiPRodri committed:
feat: add ThreatScore compliance models

AdriiiPRodri committed:
fix: update django to 5.1.8 to fix a vulnerability

AdriiiPRodri committed:
chore: update changelog

jfagoagas reviewed (changes requested):
What a great job @AdriiiPRodri 👏 

Some things that we need to review:
- Review API version, if we are creating a new endpoint we should bump for a new minor.
- We need to add tests to cover changes in both existing and new code.

Regarding some parts of the API I need @vicferpoy to review the code.

jfagoagas commented on api/src/backend/api/v1/views.py:
```diff
@@ -1291,28 +1334,97 @@ def report(self, request, pk=None):
-            try:
-                file_path = zip_files[0]
-            except IndexError as e:
-                sentry_sdk.capture_exception(e)
```
Why did you remove this?

AdriiiPRodri replied:
Because I made a refactor to avoid code repetition, now we check by using suffixes and patterns if the file exists or not:

```
keys = [obj["Key"] for obj in contents if obj["Key"].endswith(suffix)]
if not keys:
    return Response(
        {
            "detail": f"No compliance file found for name '{os.path.splitext(suffix)[0]}'."
        },
        status=status.HTTP_404_NOT_FOUND,
    )
```

jfagoagas replied:
In that case, do we need to keep `sentry_sdk.capture_exception`?

AdriiiPRodri commented:
> What a great job @AdriiiPRodri 👏
> 
> Some things that we need to review:
> 
> * Review API version, if we are creating a new endpoint we should bump for a new minor.
> * We need to add tests to cover changes in both existing and new code.
> 
> Regarding some parts of the API I need @vicferpoy to review the code.

There is a lack of tests prior to this PR, so tests have been added to the modules that previously had them. Adding new test files may take more time and as I say it is something that was missing previously. Then we will have to decide whether to create those tests in this or another PR

jfagoagas commented on api/src/backend/tasks/tasks.py:
```diff
@@ -251,84 +256,110 @@ def generate_outputs(scan_id: str, provider_id: str, tenant_id: str):
+            writer.batch_write_data_to_file(**extra)
+            writer._data.clear()
+
+        # Compliance CSVs
+        for name in frameworks_avail:
+            compliance_obj = frameworks_bulk[name]
+            klass = next(
+                (
+                    c
+                    for cond, c in COMPLIANCE_CLASS_MAP.get(provider_type, [])
+                    if cond(name)
+                ),
+                GenericCompliance,
+            )
+            filename = f"{comp_dir}_{name}.csv"
```
Again, to me this is really hard to follow.

AdriiiPRodri replied:
In order to simplify this implies more code and more testing, the MAP with the compliance is a dictionary whose key is the provider_type and inside a list of tuples, where the first element is the condition and the second one is the class

Then what we do here is to iterate that list and check the condition, if the condition is True we use the class, if none is True we use the Generic. I think that the formatting in this case of ruff makes it difficult to read

AdriiiPRodri replied:
If you want we can change to this version, it's the same:

```
for condition, cls in COMPLIANCE_CLASS_MAP.get(provider_type, []):
        if condition(framework_name):
            return cls
    return GenericCompliance
```

jfagoagas replied:
To me @vicferpoy has the final word on this PR.

vicferpoy replied:
I believe Adri already changed this. I agree it's better this way. We're getting older and I would have had issues trying to understand that piece of code in a few months 🤣

jfagoagas replied:
Totally agree, I'm getting older.

AdriiiPRodri commented on api/src/backend/tasks/tasks.py:
```diff
@@ -251,84 +256,110 @@ def generate_outputs(scan_id: str, provider_id: str, tenant_id: str):
+                ),
+                GenericCompliance,
+            )
+            filename = f"{comp_dir}_{name}.csv"
```
Yes, has the same name:

API:
`prowler-output-106908755756-20250505085215_ffiec_aws.csv`

CLI:
`prowler-output-106908755756-20250425131030_ffiec_aws.csv`

AdriiiPRodri commented on api/src/backend/tasks/tasks.py:
```diff
@@ -251,84 +256,110 @@ def generate_outputs(scan_id: str, provider_id: str, tenant_id: str):
     )
 
-    # Define auxiliary variables
+    def get_writer(writer_map, name, factory, is_last):
+        """
+        Return existing writer_map[name] or create via factory().
+        In both cases set `.close_file = is_last`.
+        """
+        initialization = False
+        if name not in writer_map:
+            writer_map[name] = factory()
+            initialization = True
+        w = writer_map[name]
+        w.close_file = is_last
+
+        return w, initialization
```
This function is to reduce as much as possible the code duplication. This function is in charge of return the correct writer and handled the initialisation of them

AdriiiPRodri commented on api/src/backend/tasks/jobs/export.py:
```diff
@@ -154,3 +244,41 @@ def _generate_output_directory(
     return path
+
+
+def _generate_compliance_output_directory(
```
It was separated to see more clearly what each one does but merging them into one can be done and I will explain how to use it in the docstring

AdriiiPRodri commented on api/src/backend/api/v1/views.py:
```diff
@@ -1219,70 +1246,86 @@ def partial_update(self, request, *args, **kwargs):
+            },
+        )
+
+    def _load_file(self, path_pattern, s3=False, bucket=None, list_objects=False):
```
This type of functions have been written to avoid duplicating code between endpoints and to delegate to each function only one specific action, so we can encapsulate the different actions

The functions returns content and filename but if something related with the file handling happens it will return it as this function is in charge of file handling

For the tests, more cases can be mocked, the function is being tested with the endpoints as part of them since they are a set. It is possible to make a specific test and mock the rest but I did not see it necessary since they are functions created specifically for the endpoints and to encapsulate actions

jfagoagas replied:
Thanks for the clarification.

AdriiiPRodri commented on api/src/backend/api/v1/views.py:
```diff
@@ -1219,70 +1246,86 @@ def partial_update(self, request, *args, **kwargs):
+                try:
+                    resp = client.list_objects_v2(Bucket=bucket, Prefix=prefix)
+                except ClientError:
+                    return Response(
```
I set a 500 because at that point, if a ClientError really appears, the error is not on the part of the user, it is something internal to us. In my opinion the errors of type 400 are a problem of configuration or in the request, in short, a user problem. The errors 500 are our errors and should never happen but if they happen in this case it is controlled. We can discuss this and change it as we decide

AdriiiPRodri replied:
I will change it but in my opinion even if boto fails it is still not the user's fault and in any case it is the server that has failed. At that point there is no way the user could have caused the crash

AdriiiPRodri replied:
What I mean is that at this point all errors are unexpected

AdriiiPRodri replied:
Change applied

jfagoagas replied:
Thanks, I know that we need to improve the way we handle this things, but I'm not sure about it.

AdriiiPRodri commented on api/pyproject.toml:
```diff
@@ -7,7 +7,7 @@ authors = [{name = "Prowler Engineering", email = "engineering@prowler.com"}]
   "celery[pytest] (>=5.4.0,<6.0.0)",
   "dj-rest-auth[with_social,jwt] (==7.0.1)",
-  "django==5.1.7",
+  "django==5.1.8",
```
Since it is very simple and this is easy to add and does not involve change of minor I thought that the quickest and easiest way is to enter here, it is to save time, if you see it convenient another PR will be created

jfagoagas replied:
If we are sure that is thoroughly tested go ahead, but for the long-term and if something happens it is best to use an specific PR for this change. What's your opinion @vicferpoy?

vicferpoy replied:
Agree @jfagoagas . But as you said, if it's already tested let's leave it here to not delay this development more. Let's keep it in mind in the future.

AdriiiPRodri commented on api/CHANGELOG.md:
```diff
@@ -8,6 +8,8 @@ All notable changes to the **Prowler API** are documented in this file.
 ### Added
 
 - Added M365 as a new provider [(#7563)](https://github.com/prowler-cloud/prowler/pull/7563).
+- Added compliance folder to export outputs as ZIP functionality [(#7653)](https://github.com/prowler-cloud/prowler/pull/7653).
+- Added a new endpoint to download the specified compliance file [(#7653)](https://github.com/prowler-cloud/prowler/pull/7653).
```
Because they are two different things, on the one hand the compliance folder has been added to the ZIP and on the other hand an endpoint has been added, but both functionalities can exist without the need of the other one. This was a last minute requirement and compliance development would not be complete without it

jfagoagas replied:
Thanks for the clarification.

AdriiiPRodri committed:
ref: merge generate folder functions

jfagoagas commented on api/CHANGELOG.md:
```diff
@@ -8,6 +8,8 @@ All notable changes to the **Prowler API** are documented in this file.
 ### Added
 
 - Added M365 as a new provider [(#7563)](https://github.com/prowler-cloud/prowler/pull/7563).
+- Added compliance folder to export outputs as ZIP functionality [(#7653)](https://github.com/prowler-cloud/prowler/pull/7653).
+- Added a new endpoint to download the specified compliance file [(#7653)](https://github.com/prowler-cloud/prowler/pull/7653).
```
```suggestion
- Added a `compliance/` folder and ZIP‐export functionality for all compliance reports.[(#7653)](https://github.com/prowler-cloud/prowler/pull/7653).
- Added a new API endpoint to fetch and download any specific compliance file by name [(#7653)](https://github.com/prowler-cloud/prowler/pull/7653).
```

Just a suggestion.

AdriiiPRodri committed:
chore: Update api/CHANGELOG.md

Co-authored-by: Pepe Fagoaga <pepe@prowler.com>

AdriiiPRodri committed:
test: modify unittests to check file_extension change

AdriiiPRodri committed:
test: add generate_outputs unittests

AdriiiPRodri committed:
ref: send to sentry

AdriiiPRodri committed:
chore: apply ruff

AdriiiPRodri committed:
ref: change the way to obtain the compliance class

AdriiiPRodri committed:
test: add tests to cover missing lines

AdriiiPRodri committed:
chore: build images based on current branch

AdriiiPRodri committed:
chore: update poetry lock

AdriiiPRodri committed:
chore: update docstrings

AdriiiPRodri committed:
test: add export unittests

vicferpoy commented on api/src/backend/api/v1/views.py:
```diff
@@ -1150,6 +1152,28 @@ def destroy(self, request, *args, pk=None, **kwargs):
+        tags=["Scan"],
+        summary="Retrieve compliance report as CSV",
+        description="Download a specific compliance report (e.g., 'cis_1.4_aws') as a CSV file.",
+        operation_id="scan_compliance_download",
```
```python
operation_id="scan_compliance_download",
```

Are we also adding this field from now on? To keep it in mind. I believe it just serves a purpose for documentation.

AdriiiPRodri replied:
For the moment I remove it and if we decide to add it we will do it at the same time in a PR

vicferpoy commented on api/src/backend/tasks/tasks.py:
```diff
@@ -251,84 +255,106 @@ def generate_outputs(scan_id: str, provider_id: str, tenant_id: str):
 
-    # Initialize the prowler provider
-    prowler_provider = initialize_prowler_provider(Provider.objects.get(id=provider_id))
+    provider_obj = Provider.objects.get(id=provider_id)
```
Could we take this opportunity to wrap this code out of the task logic itself and move it to `export.py`? To improve readability throughout the defined tasks and also allow seamless migrations to another Task Runner framework if required in the future.

I know there are old tasks that should be migrated to that pattern as well, just asking.

AdriiiPRodri replied:
This implies a lot of changes in the tests and since we know that this module has to be refactored I think it is more appropriate to do it all in a PR. The tests added in this PR have been done because we have no date for this refactor

jfagoagas replied:
Could you create a ticket for it?

vicferpoy reviewed (changes requested):
Great job with this. I left a few suggestions. Up to you to follow or discard them.

AdriiiPRodri committed:
ref: change the way to obtain the available compliance frameworks

AdriiiPRodri commented:
I can't respond directly in the comment: https://github.com/prowler-cloud/prowler/pull/7653#discussion_r2075142943

I checked it, there is nothing faster than calling `get_available_compliance_frameworks`, moved this logic to this file and implemented a simple cache to only do this calculation once per provider.

jfagoagas commented on api/src/backend/tasks/jobs/export.py:
```diff
@@ -13,13 +13,86 @@
+    "m365": [
+        (lambda name: name.startswith("cis_"), M365CIS),
+    ],
+    "nhn": [
+        (lambda name: name.startswith("iso27001_"), NHNISO27001),
+    ],
```
Please remove `nhn` because it is not an official supported provider, so it is not present in the API.

AdriiiPRodri replied:
Removed

jfagoagas commented on api/src/backend/tasks/tasks.py:
```diff
@@ -251,84 +255,106 @@ def generate_outputs(scan_id: str, provider_id: str, tenant_id: str):
         try:
-            rmtree(Path(output_directory).parent, ignore_errors=True)
-        except FileNotFoundError as e:
+            rmtree(Path(compressed).parent, ignore_errors=True)
```
Are we testing the files/path deletion? I'm not sure about but probably we are, just to double check.

AdriiiPRodri replied:
Yes, we are testing it, I change the test to be more specified

AdriiiPRodri committed:
ref: use the new function

AdriiiPRodri commented on api/src/backend/api/v1/views.py:
```diff
@@ -1219,70 +1247,109 @@ def partial_update(self, request, *args, **kwargs):
+        )
 
-        if scan_instance.output_location.startswith("s3://"):
+    def _load_file(self, path_pattern, s3=False, bucket=None, list_objects=False):
```
Right now it takes time to change this because exception responses are also handled, so they are also common to both endpoints. I didn't know the point that in Django having helper functions inside the view was a bad practice, the simplest and less time consuming solution is to repeat some code, other solutions will involve more testing and test changes

jfagoagas replied:
Could you create a ticket for it?

AdriiiPRodri committed:
fix: change function

AdriiiPRodri committed:
chore: remove nhn provider

AdriiiPRodri committed:
test: increase the test detail

AdriiiPRodri committed:
ref: change http code

jfagoagas commented on api/src/backend/api/v1/views.py:
```diff
@@ -150,6 +150,7 @@
     UserSerializer,
     UserUpdateSerializer,
 )
+from prowler.config.config import get_available_compliance_frameworks
```
Is this already reviewed?

jfagoagas reviewed (approved):
Finally 🥇 Thank you so much for the effort 👏
